### ISSUE TITLE
[FEAT]: Knowledge Graphs Support with Anything-LLM 

### ISSUE THREAD

**Issue by ishaansharma** [Mintplex-Labs/anything-llm#1008]
### What would you like to see?

I would like to see the support for adding Knowledge Graph with RAG . 
something similar to [this ](https://docs.llamaindex.ai/en/stable/examples/query_engine/knowledge_graph_rag_query_engine/).

**Comment by timothycarambat** [comment]
Being experimented in a branch called [knowledge-graph-support](https://github.com/Mintplex-Labs/anything-llm/tree/knowledge-graph-support) and it works on topic expansion and helps make answers more reliable, but I'm not sure how reliable this is when you start to use OSS LLMs

**Comment by ishaansharma** [comment]
That is great , I will be looking forward to try it when this gets merged to the main branch.

**Comment by ishaansharma** [comment]
Also can we integrate this with neo4j .

**Comment by timothycarambat** [comment]
The graph in that branch runs totally locally at this time. I decided to stay away from neo4J since it doesn't run locally and I didn't want to force users to sub or set up neo4j locally since that can be quite complex.

I understand neo4j comes with a lot of convenience though like the UI and graph browser.

**Comment by KenLau** [comment]
First of all, thanks for the contribution to the knowledge graph attempt, we have an internal project which might rely on AnythingLLM for document understanding, we would like to try knowledge graph for improve the correctness of the queries, is there any documentation or guideline which can guide us to setup the knowledge graph branch locally? Thank you very much. :D

**Comment by ishaansharma** [comment]
@timothycarambat we should integrate https://github.com/microsoft/graphrag.git to this and see the results for the same.

**Comment by StrikeNP** [comment]
I am interested in GraphRAG support as well. The last commit to the knowledge-graph-support commit was 6 months ago, but this thread was last updated 6 weeks ago. Is this feature still being developed?

**Comment by InventoCasa** [comment]
+1 for Knowledge Graph / GraphRAG support

**Comment by PSpiel** [comment]
+1 for knowledge graph support!

**Comment by realsammyt** [comment]
would love some neo4j in ALLM

**Comment by tanandy77** [comment]
> @timothycarambat we should integrate https://github.com/microsoft/graphrag.git to this and see the results for the same. 

Yes please. I would love to see this integration. Microsoft GraphRAG + AnythingLLM.

**Comment by dbcandle** [comment]
Any updates on Knowledge Graph Support?  While I personally like Neo4J, anything would be better than nothing.

Having a graph to A) be able to view as things evolve and B) provide semantic context to the LLMs would be a real game-changer!  Without something like this, everything about LLMs is a "black box", and the vector databases get messy/unwieldy over time.

**Comment by dbcandle** [comment]
> Any updates on Knowledge Graph Support? While I personally like Neo4J, anything would be better than nothing.
> 
> Having a graph to A) be able to view as things evolve and B) provide semantic context to the LLMs would be a real game-changer! Without something like this, everything about LLMs is a "black box", and the vector databases get messy/unwieldy over time (and unlike a graph, cannot be readily viewed).

In thinking...
...would this best be facilitated via MCP?  It looks like there are at least 2 Neo4J MCP servers already.

**Comment by lvca** [comment]
Coming in late on this thread with the usual disclosure: I maintain [ArcadeDB](https://github.com/ArcadeData/arcadedb), so factor that in.

@timothycarambat, your 2024 reasoning in this thread holds up well. You said *"I decided to stay away from neo4J since it doesn't run locally and I didn't want to force users to sub or set up neo4j locally since that can be quite complex."* That's exactly the bar a KG backend for AnythingLLM should meet, and it's the gap worth naming: in the OSS graph DB space today, the options that actually satisfy *local-first, no subscription, not complex to install* are a short list.

What ArcadeDB offers for this specific problem:

- **Runs locally.** Single Java jar or a 50MB Docker image. No cloud account, no API keys, no telemetry. Same operational profile users already accept for ChromaDB/LanceDB in AnythingLLM today.
- **Apache 2.0.** No subscription, no BSL, no Community-vs-Enterprise split.
- **Multi-model in one engine.** Graph + document + key/value + vector + time series against the same data. Concretely: the vectors AnythingLLM already stores could live alongside the graph, enabling the hybrid retrieval pattern (vector search to find entry nodes, then graph traversal to expand context) that usually makes GraphRAG work.
- **Graph browser UI.** Bundled web UI with a visual graph browser addresses the "convenience" point you credited Neo4j for. Runs on the same local port.
- **Bolt-compatible.** If someone later wants to plug a Neo4j-compatible library (Cypher client, langchain integration, etc.) the driver works unchanged. Also means the existing `knowledge-graph-support` branch could be adapted without a ground-up rewrite, depending on how it stores state.

To @dbcandle's MCP question from April: yes, that works. ArcadeDB ships an MCP server ([PulseMCP listing](https://www.pulsemcp.com/servers/arcadedb-arcadedb)), so AnythingLLM could talk to a local ArcadeDB over MCP rather than through a direct driver if that suits the architecture better. Both routes are fine.

Not trying to sell it here, just pointing out that the "runs locally" objection that parked Neo4j in 2024 is exactly what ArcadeDB optimizes for. If there's interest in picking the `knowledge-graph-support` branch back up, happy to help sketch what an ArcadeDB-backed version looks like, or contribute a PR once the shape of the KG integration is decided.
